platform/stm32f3_agents/cmds/transmitter.c

```diff
@@ -6,45 +6,227 @@
+	int t = 0x16FF / UART_NUM;
+	while(t--)
+		schedule();
+}
+
+static double get_data(void) {
+	double res = current_state[UART_NUM];
+	int cnt = 1;
+	int i;
+
+	for (i = 0; i < UART_NUM; i++)
+		if (current_state[i] > 0) {
+			res += current_state[i];
+			cnt++;
+		}
+
+	return res / cnt;
+}
+
+static int obtain_data(void) {
+	return leds_cnt;
+}
+
+static void transmit_data(int data, USART_TypeDef *uart) {
+	while ((STM32_USART_FLAGS(uart) & USART_FLAG_TXE) == 0);
+
+	STM32_USART_TXDATA(uart) = (uint8_t) (data + 0);
+}
+
+static void *transmitter_thread_run(void *arg) {
+	int i;
+	int data;
 
-/* Receive char via uart and turn on LED */
-void *led_handler(void *arg) {
-	uint8_t ch = 0;
 	while (1) {
-		/* Read port */
-		ch = diag_getc();
-		ch %= 8;
-		BSP_LED_Init(ch);
-		BSP_LED_On(ch);
-		ksleep(50);
-		BSP_LED_Off(ch);
-		ksleep(50);
+		data = obtain_data();
+		for (i = 0; i < UART_NUM; i++) {
+			transmit_data(0xFF, uart_base[i]);
+			transmit_delay();
+			transmit_data(data, uart_base[i]);
+			transmit_delay();
+		}
 	}
+
 	return NULL;
 }
 
-int main() {
-	uint8_t r = 0;
-	diag_init();
-	BSP_PB_Init(0, 0);
-	new_task(0, led_handler, NULL);
+static int from_neighbour(char *msg) {
+	return 1;
+}
+
+static void process_message(char *msg) {
+	if (from_neighbour(msg)) {
+		mutex_lock(&led_mutex);
+		current_state[msg[0]] = msg[1];
+		if (get_data() > leds_cnt)
+			leds_next();
+		else if (get_data() < leds_cnt)
+			leds_prev();
+		mutex_unlock(&led_mutex);
+	} else {
+		transmit_data(msg[0], uart_base[0]);
+		transmit_delay();
+		transmit_data(msg[1], uart_base[0]);
+		transmit_delay();
+	}
+}
+
+static int message_valid(char *msg) {
+	return 1;
+}
+
+static void *receiver_thread_run(void *arg) {
+	int i;
+	UART_HandleTypeDef UartHandle;
+	UartHandle.Instance = uart_base[0];
+	char buf[UART_NUM][MSG_LEN];
+	int counter[UART_NUM] = { 0 };
+	
+	int tt = 0;
+
+	mutex_init(&led_mutex);
 
 	while (1) {
-		int state = BSP_PB_GetState(0);
-		if (state) {
-			r++;
-			r %= 8;
+		tt++;
+		if (tt % 0x180 == 0) {
+			tt = 1;
+			if (BSP_PB_GetState(0))
+				leds_next();
 		}
-		ksleep(100);
-		diag_putc(r);
-		ksleep(100);
+
+		for (i = 0; i < UART_NUM; i++) {
+			if (STM32_USART_FLAGS(uart_base[i]) & USART_FLAG_RXNE) {
+				buf[i][counter[i]++] = STM32_USART_RXDATA(uart_base[i]) & 0xFF;
+				if (buf[i][counter[i] - 1] == 0xFF) {
+					counter[i] = 1;
+					continue;
+				}
+				if (counter[i] == MSG_LEN) {
+					counter[i] = 0;
+					/* Message finished */
+					if (!message_valid(buf[i])) {
+						counter[i] = 0;
+						memset(buf[i], 0, sizeof(buf[i]));
+						continue;
+					} else {
+						buf[i][0] = i;
+						process_message(buf[i]);
+					}
+				}
+			} else {
+				__HAL_USART_CLEAR_NEFLAG(&UartHandle);
+				__HAL_USART_CLEAR_FEFLAG(&UartHandle);
+				__HAL_USART_CLEAR_OREFLAG(&UartHandle);
+				//uart_base[i]->ICR = 0x0;
+				schedule();
+			}
+		}
+	}
+
+	return NULL;
+}
+
+extern void HAL_UART_MspInit(UART_HandleTypeDef *);
```

Comment:
Do you realy want to use platform specific code in a user command? Can you use ttyS instead?